internal/tokenclient/tokenclient.go

```diff
@@ -0,0 +1,146 @@
+
+package tokenclient
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/utils/clock"
+
+	"go.pinniped.dev/internal/backoff"
+	"go.pinniped.dev/internal/plog"
+)
+
+type WhatToDoWithTokenFunc func(authenticationv1.TokenRequestStatus, metav1.Duration) error
+
+type TokenClient struct {
+	namespace          string
+	serviceAccountName string
+	k8sClient          kubernetes.Interface
+	whatToDoWithToken  WhatToDoWithTokenFunc
+	expirationSeconds  int64
+	clock              clock.Clock
+	logger             plog.Logger
+}
+
+type Opt func(client *TokenClient)
+
+func WithExpirationSeconds(expirationSeconds int64) Opt {
+	return func(client *TokenClient) {
+		client.expirationSeconds = expirationSeconds
+	}
+}
+
+func New(
+	namespace string,
+	serviceAccountName string,
+	k8sClient kubernetes.Interface,
+	whatToDoWithToken WhatToDoWithTokenFunc,
+	logger plog.Logger,
+	opts ...Opt,
+) *TokenClient {
+	client := &TokenClient{
+		namespace:          namespace,
+		serviceAccountName: serviceAccountName,
+		k8sClient:          k8sClient,
+		whatToDoWithToken:  whatToDoWithToken,
+		expirationSeconds:  600,
+		clock:              clock.RealClock{},
+		logger:             logger,
+	}
+	for _, opt := range opts {
+		opt(client)
+	}
+	return client
+}
+
+type howToFetchTokenFromAPIServer func(tokenRequest *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error)
+
+func (tokenClient TokenClient) Start(ctx context.Context) {
+	sleeper := make(chan time.Time, 1)
+	// Make sure that the <-sleeper below gets run once immediately.
+	sleeper <- time.Now()
+	for {
+		select {
+		case <-ctx.Done():
+			tokenClient.logger.Info("TokenClient was cancelled and is stopping")
+			return
+		case <-sleeper:
+			var tokenTTL metav1.Duration
+			err := backoff.WithContext(ctx, &backoff.InfiniteBackoff{
+				Duration:    10 * time.Millisecond,
+				MaxDuration: 5 * time.Second,
+				Factor:      2.0,
+			}, func(ctx context.Context) (bool, error) {
+				var err error
+				var tokenRequestStatus authenticationv1.TokenRequestStatus
+				tokenRequestStatus, tokenTTL, err = tokenClient.fetchToken(
+					func(tokenRequest *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error) {
+						return tokenClient.k8sClient.CoreV1().ServiceAccounts(tokenClient.namespace).CreateToken(
+							ctx,
+							tokenClient.serviceAccountName,
+							tokenRequest,
+							metav1.CreateOptions{})
+					},
+				)
+
+				if err != nil {
+					tokenClient.logger.Warning(fmt.Sprintf("Could not fetch token: %s\n", err))
+					// We got an error. Swallow it and ask for retry.
+					return false, nil
+				}
+
+				err = tokenClient.whatToDoWithToken(tokenRequestStatus, tokenTTL)
+				if err != nil {
+					tokenClient.logger.Warning(fmt.Sprintf("unable to pass token to the caller: %s\n", err))
```

Comment:
Can all these log statements use `plog` to make them standard key/value logs?

Reply:
These log statements appear like other log statements from controllers. Do you have examples to follow?

Reply:
I updated these statements to use key/value arguments instead of Sprintf.